### PR TITLE
optimizations in int_vector_buffer

### PR DESCRIPTION
**Pull request by karasikov**

### CODE REVIEW

File: include/sdsl/construct_sa_se.hpp [hunk 1]
```diff
@@ -329,7 +329,6 @@ void _construct_sa_se(int_vector_type& text, std::string filename_sa, uint64_t s
             bkt_s[i] = bkt_lms[i];
         }
     }
-    right.buffersize(0);
```
Comment (hmusta) [inline include/sdsl/construct_sa_se.hpp:332]: From what I see, this line is meant to write the current block to disk before resetting the pointer. Does any of the subsequent code rely on this last block being written?

File: include/sdsl/construct_sa_se.hpp [hunk 2]
```diff
@@ -434,7 +433,6 @@ void _construct_sa_se(int_vector_type& text, std::string filename_sa, uint64_t s
         }
     }
     util::clear(same_lms);
-    left.buffersize(0);
```
Comment (hmusta) [inline include/sdsl/construct_sa_se.hpp:437]: see above

File: include/sdsl/construct_sa_se.hpp [hunk 3]
```diff
@@ -484,15 +482,13 @@ void _construct_sa_se(int_vector_type& text, std::string filename_sa, uint64_t s
         util::clear(lms_pos_b);
         util::clear(isa_rec);
         // write to left
-        left.buffersize(buffersize);
```
Comment (hmusta) [inline include/sdsl/construct_sa_se.hpp:487]: see above

File: include/sdsl/construct_sa_se.hpp [hunk 4]
```diff
@@ -484,15 +482,13 @@ void _construct_sa_se(int_vector_type& text, std::string filename_sa, uint64_t s
...
         left_pointer = 0;
         for (; left_pointer<number_of_lms_strings; ++left_pointer) {
             left[left_pointer] = tmp_left[number_of_lms_strings-left_pointer-1];
         }
         left_pointer--;
         util::clear(tmp_left);
     } else {
-        left.buffersize(buffersize);
```
Comment (hmusta) [inline include/sdsl/construct_sa_se.hpp:495]: see above

File: include/sdsl/construct_sa_se.hpp [hunk 5]
```diff
@@ -519,7 +515,6 @@ void _construct_sa_se(int_vector_type& text, std::string filename_sa, uint64_t s
     sdsl::remove(filename_text);
 
     // Step 12 - Scan virtual array from left to right second time
-    right.buffersize(buffersize);
```
Comment (hmusta) [inline include/sdsl/construct_sa_se.hpp:522]: see above

File: include/sdsl/lcp_support_tree2.hpp [hunk 6]
```diff
@@ -85,9 +85,10 @@ class _lcp_support_tree2
         {
             m_cst = cst;
 
-            int_vector_buffer<> lcp_buf(cache_file_name(conf::KEY_LCP, config));
+            size_type buf_len = 1000000;
```
Comment (hmusta) [inline include/sdsl/lcp_support_tree2.hpp:88]: why not leave this at 1024*1024?

File: include/sdsl/lcp_support_tree2.hpp [hunk 7]
```diff
@@ -206,9 +207,6 @@ void construct_first_child_and_lf_lcp(int_vector_buffer<>& lcp_buf,
 {
     typedef int_vector<>::size_type size_type;
     const size_type M = 255;	// limit for values represented in the small LCP part
-    size_type buf_len = 1000000;
-    lcp_buf.buffersize(buf_len);
-    bwt_buf.buffersize(buf_len);
```
Comment (hmusta) [inline include/sdsl/lcp_support_tree2.hpp:211]: see above, why not 1024*1024?

File: test/int_vector_buffer_test.cpp [hunk 8]
```diff
@@ -335,19 +335,6 @@ void test_random_access(size_type width=1)
...
-    ASSERT_EQ((size_type)1000, ivb.size());
-    for (size_type i=0; i < 1000; ++i) {
-        value_type x = dice();
-        ivb[x] = x & sdsl::bits::lo_set[ivb.width()];
-    }
-    for (size_type i=0; i < ivb.size(); ++i) {
-        if (ivb[i]) {
-            ASSERT_EQ(i & sdsl::bits::lo_set[ivb.width()], (size_type)ivb[i]);
-        }
-    }
```
Comment (hmusta) [inline test/int_vector_buffer_test.cpp:350]: is this tested somewhere else? why was this removed?

File: lib/construct_lcp.cpp [hunk 9]
```diff
@@ -57,8 +58,6 @@ void construct_lcp_semi_extern_PHI(cache_config& config)
         }
     }
 
-    size_type buffer_size = 4000000; // buffer_size is a multiple of 8!
```
Comment (hmusta) [inline lib/construct_lcp.cpp:60]: why did the original unit test change the buffer size here?

File: include/sdsl/int_vector_buffer.hpp [hunk 10]
```diff
@@ -446,8 +418,8 @@ class int_vector_buffer
                 //! Postfix increment of the proxy object
                 uint64_t operator++(int)
                 {
-                    uint64_t val = static_cast<uint64_t>(*this);
-                    ++(*this);
+                    uint64_t val = *this;
+                    m_int_vector_buffer->write(m_idx, val+1);;
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp]: double semicolon

File: include/sdsl/int_vector_buffer.hpp [hunk 11]
```diff
@@ -339,30 +322,27 @@ class int_vector_buffer
...
-                    if (m_start < m_offset) { // in case of int_vector, write header and trailing zeros
-                        uint64_t size = m_size*width();
-                        m_ofile.seekp(m_start, std::ios::beg);
-                        int_vector<t_width>::write_header(size, width(), m_ofile);
-                        assert(m_ofile.good());
-                        uint64_t wb = (size+7)/8;
-                        if (wb%8) {
-                            m_ofile.seekp(m_offset+wb);
-                            assert(m_ofile.good());
-                            m_ofile.write("\0\0\0\0\0\0\0\0", 8-wb%8);
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp:353]: wouldn't removing the padding cause problems with the buffer trying to read past the end of the file when reloaded?

File: include/sdsl/int_vector_buffer.hpp [hunk 12]
```diff
@@ -339,30 +322,27 @@ class int_vector_buffer
...
-                            assert(m_ofile.good());
-                        }
-                    }
+                if (!remove_file && m_ofile->is_open()) {
+                    flush();
+                }
+                m_ifile->close();
+                assert(m_ifile->good());
+                if (m_ofile->is_open()) {
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp:342]: why not throw an exception if one tries to close a non-open file?

File: include/sdsl/int_vector_buffer.hpp [hunk 13]
```diff
@@ -266,50 +243,56 @@ class int_vector_buffer
             return (m_buffersize*width())/8;
         }
 
-        //! Set the buffersize in bytes
-        void buffersize(uint64_t buffersize)
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp:270]: off the top of my head, I can't think of a reason for supporting a dynamic buffer size, but was this breaking anything? why remove it?

File: include/sdsl/int_vector_buffer.hpp [hunk 14]
```diff
@@ -140,11 +148,12 @@ class int_vector_buffer
          *                    If true the file will be interpreted as plain array with t_width bits per integer.
          *                    In second case (is_plain==true), t_width must be 8, 16, 32 or 64.
          */
-        int_vector_buffer(const std::string filename, std::ios::openmode mode=std::ios::in, const uint64_t buffer_size=1024*1024, const uint8_t int_width=t_width, const bool is_plain=false, const uint64_t file_offset=0)
+        int_vector_buffer(const std::string filename, std::ios::openmode mode=std::ios::in, uint64_t buffer_size=1024*1024, const uint8_t int_width=t_width, const bool is_plain=false, const uint64_t file_offset=0)
         {
             m_filename = filename;
             assert(!(mode&std::ios::app));
             mode &= ~std::ios::app;
+            mode &= ~std::ios::ate;
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp:156]: why not have this is as a runtime check? what if the user wants to use `int_vector_buffer` in append mode but then this silently changes that?

File: include/sdsl/int_vector_buffer.hpp [hunk 15]
```diff
@@ -78,17 +87,19 @@ class int_vector_buffer
...
             if (m_need_to_write) {
-                m_ofile.seekp(m_offset+(m_begin*width())/8);
-                assert(m_ofile.good());
-                if (m_begin+m_buffersize >= m_size) {
-                    //last block in file
-                    uint64_t wb = ((m_size-m_begin)*width()+7)/8;
-                    m_ofile.write(reinterpret_cast<char*>(m_buffer.data()), wb);
-                } else {
-                    m_ofile.write(reinterpret_cast<char*>(m_buffer.data()), (m_buffersize*width())/8);
+                assert(!m_ofile->is_open());
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp]: why assert that the m_ofile is not open? isn't this supposed to be writing?

**Review by hmusta — CHANGES_REQUESTED** [review]

File: include/sdsl/construct_sa_se.hpp (hunk 1 above)
Comment (karasikov) [reply include/sdsl/construct_sa_se.hpp:332]: I thought they just change the buffer size. So you think this works as flush?

File: include/sdsl/lcp_support_tree2.hpp (hunk 6 above)
Comment (karasikov) [reply include/sdsl/lcp_support_tree2.hpp:88]: I used the same value as there was before when they reset the buffer size.

File: include/sdsl/lcp_support_tree2.hpp (hunk 7 above)
Comment (karasikov) [reply include/sdsl/lcp_support_tree2.hpp:211]: I don't know why they used 1M, I guess why not...

File: test/int_vector_buffer_test.cpp (hunk 8 above)
Comment (karasikov) [reply test/int_vector_buffer_test.cpp:350]: this was testing that things work after changing the buffer size. I see this feature as useless and removed it. Thus, the test became redundant as well.

File: lib/construct_lcp.cpp (hunk 9 above)
Comment (karasikov) [reply lib/construct_lcp.cpp:60]: I have no clue why the change buffer size, but this seems like a bad logic to me.
If there is a reason, this should be implemented differently.

File: include/sdsl/int_vector_buffer.hpp (hunk 10 above)
Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp]: fixed

File: include/sdsl/int_vector_buffer.hpp (hunk 11 above)
Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp:353]: I don't know, I kept it as it was before.

File: include/sdsl/int_vector_buffer.hpp (hunk 13 above)
Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp:270]: Because this looks bad to me. If one wants to change the buffer size, just reinitialize the int_vector_buffer with a different value. That's equivalent.

File: include/sdsl/int_vector_buffer.hpp (hunk 14 above)
Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp:156]: Buffer seeks to the position where you write anyway.
So this just did nothing, essentially. Just one redundant seek.

If you want to skip a certain part of file and create the buffer in the end, you need to pass non-zero `file_offset`.
`std::ios::ate` did not control this.

File: include/sdsl/int_vector_buffer.hpp (hunk 15 above)
Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp]: my bad, fixed

File: include/sdsl/construct_sa_se.hpp (hunk 1 above)
Comment (karasikov) [reply include/sdsl/construct_sa_se.hpp:332]: I'll check all these places if flush makes sense.
But here, for instance, it's right before `reset()` so obviously there is no point to flush... I believe they just wanted to save 1 MB RAM, which is a bit weird...

Comment (karasikov) [reply include/sdsl/construct_sa_se.hpp:332]: I checked everything, didn't find any case where flush would be required.

File: include/sdsl/int_vector_buffer.hpp [hunk 16]
```diff
@@ -61,13 +66,17 @@ class int_vector_buffer
...
+                    m_ifile_pos = m_offset+(m_begin*width())/8;
+                    if (!m_ifile->good())
+                        throw std::ios_base::failure("Error when seeking in int_vector_buffer");
                 }
-                assert(m_ifile.good());
+                uint64_t wb = std::min(m_buffersize*width(), (m_size-m_begin)*width()+7)/8;
+                m_ifile->read(reinterpret_cast<char*>(m_buffer.data()), wb);
+                m_ifile_pos += wb;
+                if (!m_ifile->good())
+                    throw std::ios_base::failure("Error when reading block in int_vector_buffer");
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp:79]: doesn't `good()` return false when the end of the file is reached? won't this be triggered after reading the last block in the file? or does the padding prevent that?

File: include/sdsl/int_vector_buffer.hpp [hunk 17]
```diff
@@ -155,54 +164,49 @@ class int_vector_buffer
...
+
             if (mode & std::ios::in) {
-                uint64_t size  = 0;
+                // load size and width
+                uint64_t size = 0;
                 if (is_plain) {
-                    m_ifile.seekg(0, std::ios_base::end);
-                    size = m_ifile.tellg()*8;
+                    m_ifile->seekg(0, std::ios_base::end);
+                    size = (m_ifile->tellg()-static_cast<std::ifstream::streampos>(m_start))*8;
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp]: is there a problem with just making `m_start`, `m_ifile_pos`, and `m_ofile_pos` be of type `std::(i/o)fstream::streampos` instead of having to cast here?

File: include/sdsl/int_vector_buffer.hpp [hunk 18]
```diff
@@ -266,50 +248,54 @@ class int_vector_buffer
...
         {
-            return m_ifile.good() and m_ofile.good();
+            return m_ifile->good() and (!m_ofile->is_open() || m_ofile->good());
         }
 
         //! Returns whether underlying streams are currently associated to a file
         bool is_open()
         {
-            return m_ifile.is_open() and m_ofile.is_open();
+            return m_ifile->is_open();
```
Comment (hmusta) [inline include/sdsl/int_vector_buffer.hpp:266]: why are you only checking `m_ifile` here? shouldn't it depend on the mode?

**Review by hmusta — CHANGES_REQUESTED** [review]

File: include/sdsl/int_vector_buffer.hpp (hunk 12 above)
Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp:342]: This should never happen.
It's not allowed to call `close()` for an uninitialized stream by design.
So assert is perfectly fine.
I thought about some cases when `close()` may fail (like there is an issue with the filesystem and so on), but throwing exceptions has its own disadvantages (like you can't do it in destructor), so I'm not quite sure about this and think the previously used way with asserts is ok.

File: include/sdsl/int_vector_buffer.hpp (hunk 16 above)
Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp:79]: If you read the last character, the fail bit is not set. It's set only if you try to read beyond that, which wee don't (`wb` knows exactly where to stop)

File: include/sdsl/int_vector_buffer.hpp (hunk 18 above)
Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp:266]: `m_ifile` should always be open if the stream is active.
`m_ofile`, however, may be closed even for an initialized stream (when it's open for reading only).

File: include/sdsl/int_vector_buffer.hpp (hunk 16 above)
Comment (hmusta) [reply include/sdsl/int_vector_buffer.hpp:79]: right, but `good` also returns false if the eof bit is set
https://www.cplusplus.com/reference/ios/ios/good/

Comment (karasikov) [reply include/sdsl/int_vector_buffer.hpp:79]: Yes, but it's set only when you try to read beyond the file (not a fail bit as I said before).

Comment (hmusta) [reply include/sdsl/int_vector_buffer.hpp:79]: ah, ok then. all good

**Review by hmusta — APPROVED** [review]
